src/ints/int10_modes.cpp

```diff
@@ -324,109 +327,267 @@ std::vector<VideoModeBlock> Hercules_Mode = {
-  {0x08,0x08,0x10},{0x0a,0x08,0x10},{0x0c,0x08,0x10},{0x0e,0x08,0x10},{0x10,0x08,0x10},{0x10,0x08,0x0e},{0x10,0x08,0x0c},{0x10,0x08,0x0a},
-  {0x10,0x08,0x08},{0x10,0x0a,0x08},{0x10,0x0c,0x08},{0x10,0x0e,0x08},{0x10,0x10,0x08},{0x0e,0x10,0x08},{0x0c,0x10,0x08},{0x0a,0x10,0x08},
-  {0x08,0x10,0x08},{0x08,0x10,0x0a},{0x08,0x10,0x0c},{0x08,0x10,0x0e},{0x08,0x10,0x10},{0x08,0x0e,0x10},{0x08,0x0c,0x10},{0x08,0x0a,0x10},
-  {0x0b,0x0b,0x10},{0x0c,0x0b,0x10},{0x0d,0x0b,0x10},{0x0f,0x0b,0x10},{0x10,0x0b,0x10},{0x10,0x0b,0x0f},{0x10,0x0b,0x0d},{0x10,0x0b,0x0c},
-  {0x10,0x0b,0x0b},{0x10,0x0c,0x0b},{0x10,0x0d,0x0b},{0x10,0x0f,0x0b},{0x10,0x10,0x0b},{0x0f,0x10,0x0b},{0x0d,0x10,0x0b},{0x0c,0x10,0x0b},
-  {0x0b,0x10,0x0b},{0x0b,0x10,0x0c},{0x0b,0x10,0x0d},{0x0b,0x10,0x0f},{0x0b,0x10,0x10},{0x0b,0x0f,0x10},{0x0b,0x0d,0x10},{0x0b,0x0c,0x10}
+static cga_colors_t cga_colors_agi_amigaish = { RGBEntry
+		{0x00, 0x00, 0x00}, {0x00, 0x00, 0x3f}, {0x00, 0x2a, 0x00}, {0x00, 0x2a, 0x2a},
+		{0x33, 0x00, 0x00}, {0x2f, 0x1c, 0x37}, {0x23, 0x14, 0x00}, {0x2f, 0x2f, 0x2f},
+		{0x15, 0x15, 0x15}, {0x00, 0x2f, 0x3f}, {0x00, 0x33, 0x15}, {0x15, 0x3f, 0x3f},
+		{0x3f, 0x27, 0x23}, {0x3f, 0x15, 0x3f}, {0x3b, 0x3b, 0x00}, {0x3f, 0x3f, 0x3f},
+};
+
+// C64-like colors based on the Colodore C64 palette by pepto.
+// (brightness=50, contrast=50, saturation=50; custom dark cyan and bright
+// magenta as those colours are missing from the C64 palette)
+// https://www.colodore.com/
+static cga_colors_t cga_colors_colodore_sat50 = { RGBEntry
+		{0x00, 0x00, 0x00}, {0x0b, 0x0b, 0x26}, {0x15, 0x2b, 0x13}, {0x0e, 0x22, 0x21},
+		{0x20, 0x0c, 0x0e}, {0x23, 0x0f, 0x26}, {0x23, 0x14, 0x0a}, {0x2c, 0x2c, 0x2c},
+		{0x12, 0x12, 0x12}, {0x1c, 0x1b, 0x3b}, {0x2a, 0x3f, 0x28}, {0x1d, 0x33, 0x32},
+		{0x31, 0x1b, 0x1c}, {0x35, 0x23, 0x38}, {0x3b, 0x3c, 0x1c}, {0x3f, 0x3f, 0x3f},
+};
+
+// A 20% more saturated version of the Colodore palette (saturation=60)
+static cga_colors_t cga_colors_colodore_sat60 = { RGBEntry
+		{0x00, 0x00, 0x00}, {0x0b, 0x0a, 0x2c}, {0x13, 0x2d, 0x10}, {0x0c, 0x24, 0x22},
+		{0x23, 0x0b, 0x0d}, {0x26, 0x0d, 0x29}, {0x26, 0x13, 0x08}, {0x2c, 0x2c, 0x2c},
+		{0x12, 0x12, 0x12}, {0x1b, 0x1a, 0x3f}, {0x27, 0x3f, 0x24}, {0x1a, 0x35, 0x33},
+		{0x34, 0x19, 0x1b}, {0x37, 0x22, 0x3a}, {0x3c, 0x3d, 0x17}, {0x3f, 0x3f, 0x3f},
+};
+
+// Emulation of the actual color output of an unknown Tandy monitor.
+static cga_colors_t cga_colors_tandy_warm = { RGBEntry
+		{0x03, 0x03, 0x03}, {0x00, 0x03, 0x27}, {0x02, 0x1d, 0x05}, {0x0c, 0x23, 0x27},
+		{0x2a, 0x06, 0x00}, {0x2b, 0x0c, 0x27}, {0x2c, 0x18, 0x09}, {0x2a, 0x2a, 0x2a},
+		{0x14, 0x14, 0x14}, {0x16, 0x1a, 0x3c}, {0x11, 0x2f, 0x14}, {0x10, 0x37, 0x3e},
+		{0x3f, 0x1c, 0x14}, {0x3f, 0x21, 0x3d}, {0x3d, 0x38, 0x12}, {0x3c, 0x3c, 0x3e},
 };
+// clang-format on
+
+static std::vector<cga_colors_t> cga_colors = {
+};
+
+static void init_all_palettes(const cga_colors_t cga_colors)
```

Comment:
This can all be done at compile-time :partying_face, and can also take in a reference to the color array:

``` c++
constexpr void init_all_palettes(const cga_colors_t& cga_colors)
```

Reply:
Sure, why not 🕺🏻

Reply:
It's not often to all this constexpr goodness :laughing: 

There's something really sweet knowing that the work is only being done once - and everyone running it get those operations truly as a 0-ns noop, just by opening the the program!

Reply:
Wait a minute, we can't do that. The palettes are initialised at runtime, based on the configured CGA colours:

```cpp
void INT10_SetupPalette()
{
	auto cga_colors = configure_cga_colors();
	init_all_palettes(cga_colors);
}
```

Reply:
But the user can provide completely custom colours via `cga_colors`; we need to calculate this at runtime.

Reply:
I guess I can try it and we'll see 🤷🏻 😄

Reply:
Amazingly, it compiles! Okay, this is some black magic yet unfamiliar to me, that's why I misunderstood you first 😁

Reply:
...well, a couple of compilers don't like it, so we can't do it, after all.

**Clang (macos-10.15) (x86_64)** 

```
../src/ints/int10_modes.cpp:437:16: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
constexpr void init_all_palettes(const cga_colors_t &cga_colors)
 ```

**MSVC 32-bit** 

     4>D:\a\dosbox-staging\dosbox-staging\src\ints\int10_modes.cpp(437,16): error C3615: constexpr function 'init_all_palettes' cannot result in a constant expression [D:\a\dosbox-staging\dosbox-staging\vs\dosbox.vcxproj]

**MSVC 64-bit** 

     4>D:\a\dosbox-staging\dosbox-staging\src\ints\int10_modes.cpp(437,16): error C3615: constexpr function 'init_all_palettes' cannot result in a constant expression [D:\a\dosbox-staging\dosbox-staging\vs\dosbox.vcxproj]

Reply:
Argh 😂 was worth a shot!